lib/manager/composer/artifacts.ts

```diff
@@ -20,7 +20,68 @@ import {
 } from '../../util/fs';
 import { getRepoStatus } from '../../util/git';
 import * as hostRules from '../../util/host-rules';
-import { UpdateArtifact, UpdateArtifactsResult } from '../common';
+import {
+  UpdateArtifact,
+  UpdateArtifactsConfig,
+  UpdateArtifactsResult,
+} from '../common';
+
+function getAuthJson(config: UpdateArtifactsConfig): string | null {
+  const authJson = {};
+  let credentials = hostRules.find({
+    hostType: PLATFORM_TYPE_GITHUB,
+    url: 'https://api.github.com/',
+  });
+  // istanbul ignore if
+  if (credentials?.token) {
+    authJson['github-oauth'] = {
+      'github.com': credentials.token,
+    };
+  }
+  credentials = hostRules.find({
+    hostType: PLATFORM_TYPE_GITLAB,
+    url: 'https://gitlab.com/api/v4/',
+  });
+  // istanbul ignore if
+  if (credentials?.token) {
+    authJson['gitlab-token'] = {
+      'gitlab.com': credentials.token,
+    };
+  }
+  try {
+    // istanbul ignore else
+    if (is.array(config.registryUrls)) {
+      for (const regUrl of config.registryUrls) {
+        if (regUrl) {
+          const { host } = URL.parse(regUrl);
+          const hostRule = hostRules.find({
+            hostType: datasourcePackagist.id,
+            url: regUrl,
```

Comment:
Is there any downside if we just do a lookup without specifying `url` and use all `hosttype=packagist` authentication?

Reply:
If I understand it right, in this case we need to handle multiple `HostRule` fields, e.g. `hostName`, `domainName` and `baseUrl`.

Reply:
Take a look at changed code which you can find good enough (`baseUrl` handling requires additional `try-catch` and logging).

Reply:
registryUrls might be empty or incomplete. I don't think we want to look up based on it. Let's have this change do GitHub / gitlab as we already do and then add others after.